streaming/src/main/scala/org/apache/spark/streaming/scheduler/JobGenerator.scala

```diff
@@ -123,6 +126,12 @@ class JobGenerator(jobScheduler: JobScheduler) extends Logging {
         timedOut
       }
 
+      // generate one more bacth to make sure RDD in lastJob is checkpointed.
+      if (!jobScheduler.receiverTracker.hasUnallocatedBlocks &&
+        ssc.graph.isCheckpointMissedLastTime) {
+        Thread.sleep(ssc.graph.batchDuration.milliseconds)
```

Comment:
What about if the last batch execution takes longer time than the batchDuration? Probably we need a notification with a timeout?